include/seqan3/io/sequence_file/input.hpp

```diff
@@ -644,28 +642,100 @@ class sequence_file_input
             return;
         }
 
-        assert(!format.valueless_by_exception());
-        std::visit([&] (auto & f)
+        format->read_sequence_record(*secondary_stream, record_buffer, options);
+    }
+
+    /*!\brief An abstract base class to store the selected input format.
+     *
+     * \details
+     *
+     * This abstract base class is used to store the user given input format as a type-erased base class in a
+     * std::unique_ptr. There is exactly one derived type that implements this abstract base class called
+     * seqan3::sequence_file_input::selected_input_format which is a private subclass of
+     * seqan3::sequence_file_input. It is not exposed to the user and allows to hide the implementation detail of
+     * storing a specfic format instance which is first known at runtime.
+     */
+    struct selected_input_format_base
+    {
+        /*!\name Constructors, destructor and assignment
+         * \{
+         */
+        selected_input_format_base() = default; //!< Default.
+        selected_input_format_base(selected_input_format_base const &) = default; //!< Default.
+        selected_input_format_base(selected_input_format_base &&) = default; //!< Default.
+        selected_input_format_base & operator=(selected_input_format_base const &) = default; //!< Default.
+        selected_input_format_base & operator=(selected_input_format_base &&) = default; //!< Default.
+        virtual ~selected_input_format_base() = default; //!< Virtual default.
+        //!\}
+
+        //!\brief The interface any given user input format must implement.
+        virtual void read_sequence_record(std::istream &, record_type &, sequence_file_input_options_type const &) = 0;
+    };
+
+    /*!\brief The specific selected format to read the records from.
+     *
+     * \tparam format_t The user specific format type to store.
+     *
+     * \details
+     *
+     * This class implements the format specific read operation based on the instantiated format type.
+     * The specfic format type is selected at runtime and stored through a pointer to the abstract
+     * seqan3::sequence_file_input::selected_input_format_base class. A virtual function call then ensures that the
+     * specific read record function of the selected format is invoked.
+     */
+    template <typename format_t>
+    struct selected_input_format final : public selected_input_format_base
```

Comment:
```suggestion
    struct sequence_format final : public sequence_format_base
```

or

```suggestion
    struct sequence_format final : public selected_sequence_format
```
 Other suggestions?